OAK-9874 (unit test only) - Add a (failing and @Ignored) unit test to check that queries return same result with and without index.

nfsantos opened the pull request:
The unit tests in this PR reproduces the issue in [OAK-9874](https://issues.apache.org/jira/browse/OAK-9874)

nfsantos committed:
Add a test for checking that query results are the same with and without index. This test is currently ignored because it is failing (See OAK-9874).

tihom88 commented on oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SameQueryResultsWithAndWithoutIndexTest.java:
```diff
@@ -0,0 +1,131 @@
+        return builder.build(root.getTree("/").addChild(INDEX_DEFINITIONS_NAME).addChild(idxName));
+    }
+
+    final private static List<String> testQueries = ImmutableList.of(
```
I think most of the queries should have correct results for elastic and lucene. I think we should create 2 methods:
1. one for passing queries  and enable this method. 
2. Separate method for failing queries and annotate it with ignore.
With this we will have one active test.

nfsantos replied:
Done. Please check if this is what you suggested.

tihom88 reviewed (commented):
nitpicking: May be we should add common in SameQueryResultsWithAndWithoutIndexTest, to indicate that it is a common test for both lucene and elastic.

nfsantos commented:
> nitpicking: May be we should add common in SameQueryResultsWithAndWithoutIndexTest, to indicate that it is a common test for both lucene and elastic.

@thomasmueller suggested removing the `Common` from the names of the tests. I'm fine with either option, but let's agree on one naming pattern.

nit0906 commented:
> > nitpicking: May be we should add common in SameQueryResultsWithAndWithoutIndexTest, to indicate that it is a common test for both lucene and elastic.
> 
> @thomasmueller suggested removing the `Common` from the names of the tests. I'm fine with either option, but let's agree on one naming pattern.

Given that a lot of tests are there that still need to be ported - I would suggest to have some identifier that would suggest that these tests have already been ported simply by looking at the class name itself.

nfsantos committed:
Create separate tests for queries that are passing (same results with and without index) and an ignored tests for queries that are failing.

nfsantos committed:
Merge remote-tracking branch 'origin/trunk' into OAK-9874_unit_test

nfsantos committed:
Rename test.

nfsantos commented:
> > > nitpicking: May be we should add common in SameQueryResultsWithAndWithoutIndexTest, to indicate that it is a common test for both lucene and elastic.
> > 
> > 
> > @thomasmueller suggested removing the `Common` from the names of the tests. I'm fine with either option, but let's agree on one naming pattern.
> 
> Given that a lot of tests are there that still need to be ported - I would suggest to have some identifier that would suggest that these tests have already been ported simply by looking at the class name itself.

After talking with Thomas, I have renamed the test to include Common in the name.

thomasmueller commented on oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexAndTraversalQueriesSimilarResultsCommonTest.java:
```diff
@@ -0,0 +1,66 @@
+
+    public ElasticIndexAndTraversalQueriesSimilarResultsCommonTest() {
+        indexOptions = new ElasticIndexOptions();
+        super.passingQueries = ImmutableList.of(
```
This and LuceneIndexAndTraversalQueriesSimilarResultsCommonTest seem to have the same values for passingQueries and failingQueries... so could we move this to IndexAndTraversalQueriesSimilarResultsCommonTest ? Maybe I'm missing some aspect that wouldn't allow this... or do you think this will change in the future?

nfsantos replied:
I moved these lists to the base class to remove duplication. In the future, I expect that they may be different, but then we can move the lists to the derived classes. For the time being, I agree it's better to avoid duplication.

thomasmueller commented on oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/IndexAndTraversalQueriesSimilarResultsCommonTest.java:
```diff
@@ -0,0 +1,131 @@
+
+    protected List<String> passingQueries;
+
+    protected List<String> failingQueries;
```
Nitpicking here a bit... for me the term "queriesFailed" means that there was an exception when running the query... What about "sameResultQueries" / "differentResultQueries" or "queriesWithSameResult" / "queriesWithDifferentResult"?

nfsantos replied:
Good point, I changed the names to `.queriesWitSameResults`.

nfsantos committed:
Move the lists with queries with same and different results from the Elastic/Lucene derived classes to base class, as these are currently equal, so it is not necessary for them to be duplicated.
Enabled the test for queries with different results and make it fail if not all queries produce different results.
Change names of lists of queries from correct/wrong queries to use instead different/same results, which is more descriptive of what they are.
